atom/browser/printing/print_preview_message_handler.cc

```diff
@@ -156,28 +155,46 @@ void PrintPreviewMessageHandler::PrintToPDF(
   rfh->Send(new PrintMsg_PrintPreview(rfh->GetRoutingID(), options));
 }
 
-void PrintPreviewMessageHandler::RunPrintToPDFCallback(
+scoped_refptr<atom::util::Promise> PrintPreviewMessageHandler::GetPromise(
+    int request_id) {
+  auto it = promise_map_.find(request_id);
+  DCHECK(it != promise_map_.end());
+
+  auto promise = it->second;
+  promise_map_.erase(it);
+
+  return promise;
+}
+
+void PrintPreviewMessageHandler::ResolvePromise(
     int request_id,
     scoped_refptr<base::RefCountedMemory> data_bytes) {
   DCHECK_CURRENTLY_ON(BrowserThread::UI);
 
-  v8::Isolate* isolate = v8::Isolate::GetCurrent();
-  v8::Locker locker(isolate);
+  auto promise = GetPromise(request_id);
+
+  v8::Isolate* isolate = promise->isolate();
+  mate::Locker locker(isolate);
   v8::HandleScope handle_scope(isolate);
-  if (data_bytes && data_bytes->size()) {
-    v8::Local<v8::Value> buffer =
-        node::Buffer::Copy(isolate,
-                           reinterpret_cast<const char*>(data_bytes->front()),
-                           data_bytes->size())
-            .ToLocalChecked();
-    print_to_pdf_callback_map_[request_id].Run(v8::Null(isolate), buffer);
-  } else {
-    v8::Local<v8::String> error_message =
-        v8::String::NewFromUtf8(isolate, "Failed to generate PDF");
-    print_to_pdf_callback_map_[request_id].Run(
-        v8::Exception::Error(error_message), v8::Null(isolate));
-  }
-  print_to_pdf_callback_map_.erase(request_id);
+  v8::MicrotasksScope script_scope(isolate,
```

Comment:
the util file deals with scoping prior to resolution; this can be removed

Reply:
will remove the `script_scope`, should I keep the `context_scope` though?

Reply:
we handle it in the util!
```
  v8::Maybe<bool> Resolve() {
    v8::HandleScope handle_scope(isolate());
    v8::MicrotasksScope script_scope(isolate(),
                                     v8::MicrotasksScope::kRunMicrotasks);
    v8::Context::Scope context_scope(
        v8::Local<v8::Context>::New(isolate(), GetContext()));

    return GetInner()->Resolve(GetContext(), v8::Undefined(isolate()));
  }
```

Reply:
I know, but don't we need that because we are creating the `v8::Local<v8::Value> buffer` before calling the util?